clustering/common/src/main/java/org/jboss/as/clustering/controller/ServiceValueExecutor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2019, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.controller;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.jboss.msc.service.ServiceName;
+import org.wildfly.common.function.ExceptionFunction;
+
+/**
+ * Executes a given function against the captured value of a service.
+ * @author Paul Ferraro
+ */
+public class ServiceValueExecutor<T> implements ServiceValueCaptor<T>, FunctionExecutor<T> {
+
+    private final ServiceName name;
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private volatile T value = null;
+
+    public ServiceValueExecutor(ServiceName name) {
+        this.name = name;
+    }
+
+    @Override
+    public ServiceName getServiceName() {
+        return this.name;
+    }
+
+    @Override
+    public void accept(T value) {
+        Lock lock = this.lock.writeLock();
+        lock.lock();
+        try {
+            this.value = value;
```

Comment:
If you'll eliminate the lock, accept method will contain just "this.value = value;" statement

Reply:
See write lock reasoning below.